### ISSUE TITLE
`std::thread::hardware_concurrency` return incorrect count threads in NUMA multi-CPU configuration

### ISSUE THREAD

**Issue by GermanAizek** [llvm/llvm-project#72267]
I have Xeon 2xE5-2699v4 Windows detect 88 threads and 44 cores. I noticed that all binaries that are compiled using Clang incorrectly detect count threads on multi 2x 4x socket motherboards with NUMA architecture. Therefore, I propose my own version, which is modified and works on all systems. I read on visual studio forum and realized that this is a WinAPI problem.

```cpp
#include <iostream>
#include <Windows.h>
#include <thread>

// modified my function
size_t MyHardwareConcurrency() noexcept
{
	DWORD length = 0;
	size_t concurrency = 0;
	const auto validConcurrency = [&concurrency]() noexcept -> size_t {
		if (concurrency == 0)
			return std::thread::hardware_concurrency();
		else
			return concurrency;
		};
	if (GetLogicalProcessorInformationEx(RelationAll, nullptr, &length) != FALSE) {
		return validConcurrency();
	}
	if (GetLastError() != ERROR_INSUFFICIENT_BUFFER) {
		return validConcurrency();
	}
	std::unique_ptr<void, void (*)(void*)> buffer(std::malloc(length), std::free);
	if (!buffer) {
		return validConcurrency();
	}
	auto* mem = reinterpret_cast<unsigned char*>(buffer.get());
	if (GetLogicalProcessorInformationEx(
		RelationAll, reinterpret_cast<PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX>(mem), &length) == false) {
		return validConcurrency();
	}
	DWORD i = 0;
	while (i < length) {
		const auto* proc = reinterpret_cast<PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX>(mem + i);
		if (proc->Relationship == RelationProcessorCore) {
			for (WORD group = 0; group < proc->Processor.GroupCount; ++group) {
				for (KAFFINITY mask = proc->Processor.GroupMask[group].Mask; mask != 0; mask >>= 1) {
					concurrency += mask & 1;
				}
			}
		}
		i += proc->Size;
	}
	return validConcurrency();
}

uint32_t get_thread_count()
{
    static const uint32_t g_count = []()
        {
#ifdef _WIN32
            ::SYSTEM_INFO sysInfo;
            ::GetNativeSystemInfo(&sysInfo);
            return sysInfo.dwNumberOfProcessors;
#else
            return ::sysconf(_SC_NPROCESSORS_ONLN);
#endif
        }();

        return g_count;
}

// in llvm-project/libcxx/src/thread.cpp
DWORD getWin32Count()
{
	SYSTEM_INFO info;
	GetSystemInfo(&info);
	return info.dwNumberOfProcessors;
}

int main()
{
    std::cout << MyHardwareConcurrency() << '\n';
    std::cout << get_thread_count() << '\n';
    std::cout << getWin32Count() << '\n';
}
```

Result: MSVC v19.37 latest
```
88
44
44
```

**Comment by philnik777** [comment]
Feel free to create a patch.

**Comment by GermanAizek** [comment]
@philnik777 it has already been created here: https://github.com/llvm/llvm-project/pull/72270
